### PR TITLE
Alerting: Create writer interface for recording rules

### PR DESCRIPTION
**Pull request by rwwiv**
**What is this feature?**

This PR introduces a `Writer` interface that recording rules call to write the frames for a particular refID returned from a successful evaluation. For now we pass a fake (also used in tests) as a stub implementation.

**Why do we need this feature?**

First step towards recording rules writing their metrics.

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/ngalert/schedule/recording_rule.go [hunk 1]
```diff
@@ -212,3 +240,17 @@ func (r *recordingRule) evaluationDoneTestHook(ev *Evaluation) {
...
 }
+
+func (r *recordingRule) frameRef(refID string, resp *backend.QueryDataResponse) (data.Frames, error) {
+	if len(resp.Responses) == 0 {
+		return nil, fmt.Errorf("no responses returned from rule evaluation")
+	}
+
+	for ref, resp := range resp.Responses {
+		if ref == refID {
+			return resp.DeepCopy().Frames, nil
```
Comment (alexweav) [inline pkg/services/ngalert/schedule/recording_rule.go:251]: The response object can be absolutely enormous in certain cases - megabytes even, it contains every single result of every single expression node. So, number of dimensions * number of expression nodes. Is there a way we can avoid a deep copy?

File: pkg/services/ngalert/schedule/recording_rule.go [hunk 2]
```diff
@@ -184,10 +189,33 @@ func (r *recordingRule) tryEvaluation(ctx context.Context, ev *Evaluation, logge
...
+
+	frames, err := r.frameRef(ev.rule.Record.From, result)
+	if err != nil {
+		span.SetStatus(codes.Error, "failed to extract frames from rule evaluation")
+		span.RecordError(err)
+		return fmt.Errorf("failed to extract frames from rule evaluation: %w", err)
+	}
+
+	writeStart := r.clock.Now()
+	err = r.writer.Write(ctx, ev.rule.Record.Metric, writeStart, frames)
```
Comment (alexweav) [inline pkg/services/ngalert/schedule/recording_rule.go:205]: writer will also need to be passed `rule.Labels`, the series will need to inherit the labels from the rule definition too

File: pkg/services/ngalert/schedule/recording_rule.go (hunk 1 above)
Comment (rwwiv) [reply pkg/services/ngalert/schedule/recording_rule.go:251]: Whoops that shouldn't have made it in, was using this for debugging.

File: pkg/services/ngalert/schedule/recording_rule.go (hunk 2 above)
Comment (rwwiv) [reply pkg/services/ngalert/schedule/recording_rule.go:205]: Good call

**Review by alexweav — APPROVED** [review]

**Pull request merged**